src/rsz/src/RemoveBuffer.cc

```diff
@@ -0,0 +1,169 @@
+// All rights reserved.
+//
+// BSD 3-Clause License
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "RemoveBuffer.hh"
+
+#include "rsz/Resizer.hh"
+#include "sta/Corner.hh"
+#include "sta/Graph.hh"
+#include "sta/Liberty.hh"
+#include "sta/Parasitics.hh"
+#include "sta/PortDirection.hh"
+#include "sta/Sdc.hh"
+#include "sta/Units.hh"
+#include "utl/Logger.h"
+
+namespace rsz {
+
+using std::string;
+using utl::RSZ;
+
+using sta::NetPinIterator;
+using sta::Pin;
+using sta::Port;
+
+RemoveBuffer::RemoveBuffer(Resizer* resizer) : resizer_(resizer)
+{
+}
+
+void RemoveBuffer::init()
+{
+  logger_ = resizer_->logger_;
+  sta_ = resizer_->sta_;
+  db_network_ = resizer_->db_network_;
+  copyState(sta_);
+}
+
+// There are two buffer removal modes: auto and manual:
+// 1) auto mode: this happens during setup fixing, power recovery, buffer
+// removal
+//      Dont-touch, fixed cell and boundary buffer constraints are honored
+// 2) manual mode: this happens during manual buffer removal during ECO
+//      This ignores dont-touch, fixed cell and boundary buffer constraints
+bool RemoveBuffer::removeBuffer(Instance* buffer, bool honorDontTouchFixed)
+{
+  init();
+  LibertyCell* lib_cell = network_->libertyCell(buffer);
+  if (!lib_cell || !lib_cell->isBuffer()) {
+    return false;
+  }
+  dbInst* db_inst = db_network_->staToDb(buffer);
+  if (honorDontTouchFixed
```

Comment:
If you are not honoring don't touch then you need to remove the setting before trying to delete such an instance (otherwise odb will throw an exception).